numpy/core/tests/test_longdouble.py

```diff
@@ -8,7 +8,15 @@
     )
 from numpy.core.tests._locales import CommaDecimalPointLocale
 
-LD_INFO = np.finfo(np.longdouble)
+
+def initialize_ld_info():
+    """Initialize longdouble info supressing the user warning."""
+    with warnings.catch_warnings():
+        warnings.filterwarnings('always', '', UserWarning)
+        return np.finfo(np.longdouble)
```

Comment:
Why does this require a warning context?

Reply:
Given Matti's suggestions, at the end I decided to instead of bringing up an error stick to a warning. As the tests of longdouble call to finfo, if I didn't put the warning context then all these tests will fail because the warning will be seen as an error.

Reply:
Isn't the warning giving at attribute access only?

Reply:
Yes, but the current way to construct the `finfo` class in runtime is to basically make a copy of all the `MachArLike` attributes into the new class. So it will access all the attributes while creating `finfo`. That's what I meant in my previous comment that changing this will imply a huge refactor on how these classes interact.

Reply:
Then the warning is tagged onto the wrong object, it should be on `np.finfo`, nobody should use `MaArch(Like)` directly anyway.

Reply:
I can add it on both, at the end `MaArchLike` is the one that has all the information that `finfo` needs.

Reply:
If you make sure that you don't give the warning unnecessarily, fine.  I think it just unnecessary complexity to add it there, though.

Reply:
@seberg can you please review the latest changes I added? I removed the warning from the `MaArchLike` class and put in only in the `finfo` class. These lines are still necessary for the longdouble test.

Reply:
> I removed the warning from the MaArchLike class

Then why is this still necessary?  We don't want unnecessary warnings, so it should happen exactly when you access the `.smallest_normal` or `.tiny` attribute and we had set it to "invalid".  


PS, trying in valgrind (which has broken long-double), I get:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-1-abde4ba92c1e> in <module>
----> 1 np.finfo(np.longdouble)

~/forks/numpy/build/testenv/lib/python3.9/site-packages/numpy/core/getlimits.py in __new__(cls, dtype)
    482         if obj is not None:
    483             return obj
--> 484         obj = object.__new__(cls)._init(dtype)
    485         for dt in dtypes:
    486             cls._finfo_cache[dt] = obj

~/forks/numpy/build/testenv/lib/python3.9/site-packages/numpy/core/getlimits.py in _init(self, dtype)
    509         self._str_eps = machar._str_eps.strip()
    510         self._str_resolution = machar._str_resolution.strip()
--> 511         self._str_smallest_normal = machar._str_smallest_normal.strip()
    512         self._str_smallest_subnormal = machar._str_smallest_subnormal.strip()
    513         return self

AttributeError: 'MachAr' object has no attribute '_str_smallest_normal'
```
I don't really think it matters to fix `MachAr`, but since we touched it, probably no reason not to.

Reply:
> Then why is this still necessary? We don't want unnecessary warnings, so it should happen exactly when you access the .smallest_normal or .tiny attribute and we had set it to "invalid".

That is exactly what is happening as it is now. It is not throwing unnecessary warnings.

> PS, trying in valgrind (which has broken long-double), I get:

That's a weird error because `_str_smallest_subnormal` is being defined in line 61. Did you had any other output?

Reply:
Valgrind gets into the `MachAr` path, not `MachArLike`... Probably that should never happen in normal circumstances. Valgrind just is broken for longdouble

Reply:
OK, great!  If you can delete this warning capture, then that is the important part.

Reply:
Sure, I've just removed it on the last commit and all tests are working as expected. I think this is finally ready to merge